### PR TITLE
issue #724: Model binding broken with enum keys

### PR DESCRIPTION
**Pull request by xuzhg**
### Issues
#724 

### Description
support enum key model binding

### CODE REVIEW

File: OData/src/System.Web.OData/OData/Routing/Conventions/ActionRoutingConvention.cs [hunk 1]
```diff
@@ -41,7 +41,11 @@ public override string SelectAction(ODataPath odataPath, HttpControllerContext c
                         string actionName = GetAction(odataPath).SelectAction(actionMap, isCollection: false);
                         if (actionName != null)
                         {
-                            controllerContext.AddKeyValueToRouteData(odataPath);
+                            EntitySetPathSegment entitySetPathSegment = (EntitySetPathSegment)odataPath.Segments.First();
+                            IEdmEntityType edmEntityType = entitySetPathSegment.EntitySetBase.EntityType();
+                            KeyValuePathSegment keyValueSegment = (KeyValuePathSegment)odataPath.Segments[1];
+
+                            controllerContext.AddKeyValueToRouteData(keyValueSegment, edmEntityType, ODataRouteConstants.Key);
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Routing/Conventions/ActionRoutingConvention.cs]: warp the line longer than 120

File: OData/src/System.Web.OData/OData/Routing/Conventions/ProcedureRoutingConventionHelpers.cs [hunk 2]
```diff
@@ -54,15 +56,94 @@ public static string SelectAction(this IEdmOperation operation, ILookup<string,
...
 
-            KeyValuePathSegment keyValueSegment = odataPath.Segments[1] as KeyValuePathSegment;
-            if (keyValueSegment != null)
+            IDictionary<string, object> routingConventionsStore = controllerContext.Request.ODataProperties().RoutingConventionsStore;
+
+            if (entityType == null && keySegment.Segment != null)
             {
-                controllerContext.RouteData.Values[ODataRouteConstants.Key] = keyValueSegment.Value;
+                entityType = keySegment.Segment.EdmType as IEdmEntityType;
+            }
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Routing/Conventions/ProcedureRoutingConventionHelpers.cs]: add empty line here

File: OData/src/System.Web.OData/OData/Routing/Conventions/ProcedureRoutingConventionHelpers.cs [hunk 3]
```diff
@@ -54,15 +56,94 @@ public static string SelectAction(this IEdmOperation operation, ILookup<string,
...
+                    if (alternateKey || keyCount > 1)
+                    {
+                        newKeyName = keyPrefix + keyValue.Key;
+                    }
+                    else
+                    {
+                        newKeyName = keyPrefix;
+                    }
+
+                    object value = ODataUriUtils.ConvertFromUriLiteral(keyValue.Value, ODataVersion.V4, model, keyProperty.Type);
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Routing/Conventions/ProcedureRoutingConventionHelpers.cs]: can we extract this into a method?

File: OData/src/System.Web.OData/OData/Routing/Conventions/ProcedureRoutingConventionHelpers.cs [hunk 4]
```diff
@@ -54,15 +56,94 @@ public static string SelectAction(this IEdmOperation operation, ILookup<string,
...
+                        newKeyName = keyCount == 1 ? keyPrefix : keyPrefix + keyValuePair.Key;
+                    }
+
+                    ODataPathSegmentExtenstions.AddKeyValues(newKeyName, keyValuePair.Value, keyProperty.Type,
+                        controllerContext.RouteData.Values,
+                        routingConventionsStore);
+                }
+            }
+            else
+            {
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Routing/Conventions/ProcedureRoutingConventionHelpers.cs]: can we extract this into a method?

File: OData/src/System.Web.OData/OData/Routing/Conventions/ActionRoutingConvention.cs (hunk 1 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Routing/Conventions/ActionRoutingConvention.cs]: Re-lined.

---
In reply to: [63321263](https://github.com/OData/WebApi/pull/727#discussion_r63321263) [](ancestors = 63321263)

File: OData/src/System.Web.OData/OData/Routing/Conventions/ProcedureRoutingConventionHelpers.cs (hunk 2 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Routing/Conventions/ProcedureRoutingConventionHelpers.cs]: added

---
In reply to: [63321994](https://github.com/OData/WebApi/pull/727#discussion_r63321994) [](ancestors = 63321994)

**Comment by xuzhg** [comment]
Merged. Thanks.

**Pull request closed without merging**